File: revenue-navigator/src/lib/api/analytics.ts
import { fetchApi } from './client';

export interface PortfolioKpis {
  total_accounts: number;
  churn_risk_count: number;
  avg_health_score: number;
  avg_relationship_score: number;
  avg_sentiment_score: number;
  total_revenue: number;
  upsell_pipeline: number;
  avg_utilization_percent: number;
  accounts_on_track: number;
  upsell_account_count: number;
  renewed_count: number;
}

export interface LifecycleStageMetric {
  stage: string;
  label: string;
  count: number;
  revenue: number;
}

export interface RenewalQuarterMetric {
  quarter: string;
  label: string;
  count: number;
  revenue: number;
  days_range: string;
}

export interface RiskDistributionItem {
  name: string;
  value: number;
}

export interface UpsellStageMetric {
  stage: string;
  count: number;
  value: number;
}

export interface PortfolioConsumption {
  avg_deployment_pct: number;
  products_at_risk: number;
  unused_entitlements: number;
  accounts_needing_action: number;
}

export interface ScoreBucket {
  bucket: string;
  count: number;
}

export interface ScoreDistribution {
  metric: string;
  buckets: ScoreBucket[];
}

export interface TopAtRiskAccount {
  account_id: string;
  account_name: string;
  stage: string;
  stage_label: string;
  risk_score: number;
  renewal_days: number | null;
  revenue: number;
}

export interface AnalyticsGoals {
  upsell_pipeline_target: number;
  renewal_target_percent: number;
  high_risk_threshold_percent: number;
  upsell_pipeline_actual: number;
  renewal_rate_percent: number;
  high_risk_count: number;
}

export interface PortfolioAnalyticsResponse {
  kpis: PortfolioKpis;
  lifecycle_stages: LifecycleStageMetric[];
  renewal_quarters: RenewalQuarterMetric[];
  risk_distribution: RiskDistributionItem[];
  upsell_by_stage: UpsellStageMetric[];
  portfolio_consumption: PortfolioConsumption;
  score_distributions: ScoreDistribution[];
  top_at_risk_accounts: TopAtRiskAccount[];
  goals: AnalyticsGoals;
}

export interface TrendDataPoint {
  month: string;
  total_revenue: number;
  churn_risk_count: number;
  renewed_count: number;
  at_risk_count: number;
  avg_health_score: number;
}

export interface AnalyticsTrendsResponse {
  months: number;
  series: TrendDataPoint[];
}

export interface DashboardStatsResponse {
  total_accounts: number;
  churn_risk_count: number;
  avg_health_score: number;
  avg_relationship_score: number;
  avg_sentiment_score: number;
  total_arr: number;
  upsell_pipeline: number;
}

function billingQuery(billingInterval?: 'monthly' | 'annual') {
  return billingInterval ? `?billing_interval=${billingInterval}` : '';
}

export const analyticsApi = {
  getDashboardStats: async (billingInterval?: 'monthly' | 'annual'): Promise<DashboardStatsResponse> => {
    return fetchApi(`/analytics/dashboard${billingQuery(billingInterval)}`);
  },

  getPortfolioAnalytics: async (
    billingInterval?: 'monthly' | 'annual'
  ): Promise<PortfolioAnalyticsResponse> => {
    return fetchApi(`/analytics/portfolio${billingQuery(billingInterval)}`);
  },

  getAnalyticsTrends: async (
    billingInterval?: 'monthly' | 'annual',
    months = 12
  ): Promise<AnalyticsTrendsResponse> => {
    const params = new URLSearchParams();
    if (billingInterval) params.set('billing_interval', billingInterval);
    params.set('months', String(months));
    return fetchApi(`/analytics/trends?${params.toString()}`);
  },
};

File: Backend/app/schemas/analytics.py
"""Pydantic schemas for portfolio analytics endpoints."""
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class PortfolioKpis(BaseModel):
    total_accounts: int = 0
    churn_risk_count: int = 0
    avg_health_score: float = 0.0
    avg_relationship_score: float = 0.0
    avg_sentiment_score: float = 0.0
    total_revenue: float = 0.0
    upsell_pipeline: float = 0.0
    avg_utilization_percent: float = 0.0
    accounts_on_track: int = 0
    upsell_account_count: int = 0
    renewed_count: int = 0


class LifecycleStageMetric(BaseModel):
    stage: str
    label: str
    count: int = 0
    revenue: float = 0.0


class RenewalQuarterMetric(BaseModel):
    quarter: str
    label: str
    count: int = 0
    revenue: float = 0.0
    days_range: str


class RiskDistributionItem(BaseModel):
    name: str
    value: int = 0


class UpsellStageMetric(BaseModel):
    stage: str
    count: int = 0
    value: float = 0.0


class PortfolioConsumption(BaseModel):
    avg_deployment_pct: int = 0
    products_at_risk: int = 0
    unused_entitlements: int = 0
    accounts_needing_action: int = 0


class ScoreBucket(BaseModel):
    bucket: str
    count: int = 0


class ScoreDistribution(BaseModel):
    metric: str
    buckets: List[ScoreBucket] = Field(default_factory=list)


class TopAtRiskAccount(BaseModel):
    account_id: str
    account_name: str
    stage: str
    stage_label: str
    risk_score: float = 0.0
    renewal_days: Optional[int] = None
    revenue: float = 0.0


class AnalyticsGoals(BaseModel):
    upsell_pipeline_target: float = 0.0
    renewal_target_percent: float = 0.0
    high_risk_threshold_percent: float = 70.0
    upsell_pipeline_actual: float = 0.0
    renewal_rate_percent: float = 0.0
    high_risk_count: int = 0


class PortfolioAnalyticsResponse(BaseModel):
    kpis: PortfolioKpis
    lifecycle_stages: List[LifecycleStageMetric] = Field(default_factory=list)
    renewal_quarters: List[RenewalQuarterMetric] = Field(default_factory=list)
    risk_distribution: List[RiskDistributionItem] = Field(default_factory=list)
    upsell_by_stage: List[UpsellStageMetric] = Field(default_factory=list)
    portfolio_consumption: PortfolioConsumption = Field(default_factory=PortfolioConsumption)
    score_distributions: List[ScoreDistribution] = Field(default_factory=list)
    top_at_risk_accounts: List[TopAtRiskAccount] = Field(default_factory=list)
    goals: AnalyticsGoals = Field(default_factory=AnalyticsGoals)


class TrendDataPoint(BaseModel):
    month: str
    total_revenue: float = 0.0
    churn_risk_count: int = 0
    renewed_count: int = 0
    at_risk_count: int = 0
    avg_health_score: float = 0.0


class AnalyticsTrendsResponse(BaseModel):
    months: int = 12
    series: List[TrendDataPoint] = Field(default_factory=list)
